src/components/interpolatedColorLegend.ts

```diff
@@ -0,0 +1,249 @@
+      var lowerWriteOptions = {
+                selection: this._lowerLabel,
+                xAlign: "center",
+                yAlign: "center",
+                textRotation: 0
+            };
+      var upperWriteOptions = {
+                selection: this._upperLabel,
+                xAlign: "center",
+                yAlign: "center",
+                textRotation: 0
+            };
+
+      var swatchWidth: number;
+      var swatchHeight: number;
+      var swatchX: (d: any, i: number) => number;
+      var swatchY: (d: any, i: number) => number;
+
+      if (this._isVertical()) {
+        var longestTextWidth = text1Width > text0Width ? text1Width : text0Width;
+        swatchWidth = this.width() - 3 * padding - longestTextWidth;
+        swatchHeight = (this.height() - 2 * padding) / numSwatches;
+        swatchY = (d: any, i: number) => padding + (numSwatches - (i + 1)) * swatchHeight;
+
+        upperWriteOptions.yAlign = "top";
+        upperLabelShift.y = padding;
+        lowerWriteOptions.yAlign = "bottom";
+        lowerLabelShift.y = -padding;
+
+        if (this._orientation === "left") {
+          swatchX = (d: any, i: number) => padding + longestTextWidth + padding;
+          upperWriteOptions.xAlign = "right";
+          upperLabelShift.x = -(padding + swatchWidth + padding);
+          lowerWriteOptions.xAlign = "right";
+          lowerLabelShift.x = -(padding + swatchWidth + padding);
+        } else { // right
+          swatchX = (d: any, i: number) => padding;
+          upperWriteOptions.xAlign = "left";
+          upperLabelShift.x = padding + swatchWidth + padding;
+          lowerWriteOptions.xAlign = "left";
+          lowerLabelShift.x = padding + swatchWidth + padding;
+        }
+      } else { // horizontal
+        swatchWidth = (this.width() - 4 * padding - text0Width - text1Width) / numSwatches;
+        swatchHeight = this.height() - 2 * padding;
+        swatchX = (d: any, i: number) => (padding + text0Width + padding) + i * swatchWidth;
+        swatchY = (d: any, i: number) => padding;
+
+        upperWriteOptions.xAlign = "right";
+        upperLabelShift.x = -padding;
+        lowerWriteOptions.xAlign = "left";
+        lowerLabelShift.x = padding;
+      }
+
+      this._upperLabel.text(""); // clear the upper label
+      this._writer.write(text1, this.width(), this.height(), upperWriteOptions);
+      var upperTranslateString = "translate(" + upperLabelShift.x + ", " + upperLabelShift.y + ")";
+      this._upperLabel.attr("transform", upperTranslateString);
+
+      this._lowerLabel.text(""); // clear the lower label
+      this._writer.write(text0, this.width(), this.height(), lowerWriteOptions);
+      var lowerTranslateString = "translate(" + lowerLabelShift.x + ", " + lowerLabelShift.y + ")";
+      this._lowerLabel.attr("transform", lowerTranslateString);
+
+      swatchWidth = Math.max(swatchWidth, 0);
+      swatchHeight = Math.max(swatchHeight, 0);
```

Comment:
This correcting shoulnd't be done here.  Mightas well do it where all the other calculations are being done.

Also, Math.abs